text/0000-stabilize-no_std.md

```diff
@@ -0,0 +1,195 @@
+worry about it.
+
+# Drawbacks
+
+The current distribution provides precisely one library, the standard library,
+for general consumption of Rust programs. Adding a new one (libcore) is adding
+more surface area to the distribution (in addition to adding a new `#![no_core]`
+attribute). This surface area is greatly desired, however.
+
+When using `#![no_std]` the experience of Rust programs isn't always the best as
+there are some pitfalls that can be run into easily. For example, macros and
+plugins sometimes hardcode `::std` paths, but most ones in the standard
+distribution have been updated to use `::core` in the case that `#![no_std]` is
+present. Another example is that common utilities like vectors, pointers, and
+owned strings are not available without liballoc, which will remain an unstable
+library. This means that users of `#![no_std]` will have to reimplement all of
+this functionality themselves.
+
+This RFC does not yet pave a way forward for using `#![no_std]` and producing an
+executable because the `#[start]` item is required, but remains feature gated.
+This RFC just enables creation of Rust static or dynamic libraries which don't
+depend on the standard library in addition to Rust libraries (rlibs) which do
+not depend on the standard library.
+
+On the topic of lang items, it's somewhat unfortunate that the implementation of
+a panic cannot be defined on stable Rust. The `#![lang_items_abort]` attribute
+unconditionally defines all lang items, including `panic_fmt`, so it's not
+possible to provide a custom implementation of the `panic_fmt` lang item while
+still asking the compiler to define others like `eh_personality` and
+`stack_exhausted`.
+
+In stabilizing the `#![no_std]` attribute it's likely that a whole ecosystem of
+crates will arise which work with `#![no_std]`, but in theory all of these
+crates should also interoperate with the rest of the ecosystem using `std`.
+Unfortunately, however, there are known cases where this is not possible. For
+example if a macro is exported from a `#![no_std]` crate which references items
+from `core` it won't work by default with a `std` library.
+
+# Alternatives
+
+Most of the strategies taken in this RFC have some minor variations on what can
+happen:
+
+* The `#![no_std]` attribute could be stabilized as-is without adding a
+  `#![no_core]` attribute, requiring users to write `extern crate core` and
+  import the core prelude manually. The burden of adding `#![no_core]` to the
+  compiler, however, is seen as not-too-bad compared to the increase in
+  ergonomics of using `#![no_std]`.
+* The lang items could not be required to be defined, and the compiler could
+  provide aborting stubs to be linked in if they aren't defined anywhere else.
+  This has the downside of perhaps silently aborting a program, however, without
+  an explicit opt-in.
+* The compiler could not require `eh_personality` or `stack_exhausted` if no
```

Comment:
Writing kernel code in Rust requires recompiling libcore with custom settings anyway.  Making rustc do this seems reasonable regardless of anything else we implement.